src/dict.c

```diff
@@ -815,242 +812,60 @@ static unsigned long rev(unsigned long v) {
- * the old bucket: for example let's say we already iterated with
- * a 4 bit cursor 1100 (the mask is 1111 because hash table size = 16).
- *
- * If the hash table will be resized to 64 elements, then the new mask will
- * be 111111. The new buckets you obtain by substituting in ??1100
- * with either 0 or 1 can be targeted only by keys we already visited
- * when scanning the bucket 1100 in the smaller hash table.
- *
- * By iterating the higher bits first, because of the inverted counter, the
- * cursor does not need to restart if the table size gets bigger. It will
- * continue iterating using cursors without '1100' at the end, and also
- * without any other combination of the final 4 bits already explored.
- *
- * Similarly when the table size shrinks over time, for example going from
- * 16 to 8, if a combination of the lower three bits (the mask for size 8
- * is 111) were already completely explored, it would not be visited again
- * because we are sure we tried, for example, both 0111 and 1111 (all the
- * variations of the higher bit) so we don't need to test it again.
- *
- * WAIT... YOU HAVE *TWO* TABLES DURING REHASHING!
- *
- * Yes, this is true, but we always iterate the smaller table first, then
- * we test all the expansions of the current cursor into the larger
- * table. For example if the current cursor is 101 and we also have a
- * larger table of size 16, we also test (0)101 and (1)101 inside the larger
- * table. This reduces the problem back to having only one table, where
- * the larger one, if it exists, is just an expansion of the smaller one.
+ * The cursor is basically a path into the hash tree. The keys are iterated in
+ * the order of how they are stored in the hash tree, in depth first order. The
+ * 5 least significant bits of the cursor are used as an index into the first
+ * level of nodes. The next 5 bits are used an index into the next level and so
+ * forth.
  *
  * LIMITATIONS
  *
  * This iterator is completely stateless, and this is a huge advantage,
  * including no additional memory used.
  *
- * The disadvantages resulting from this design are:
+ * It also has the following nice properties:
+ *
+ * 1) No duplicates. Each element is returned only once.
  *
- * 1) It is possible we return elements more than once. However this is usually
- *    easy to deal with in the application level.
- * 2) The iterator must return multiple elements per call, as it needs to always
- *    return all the keys chained in a given bucket, and all the expansions, so
- *    we are sure we don't miss keys moving during rehashing.
- * 3) The reverse cursor is somewhat hard to understand at first, but this
- *    comment is supposed to help.
+ * 2) The iterator can usually return exactly the requested number of entries.
+ *    The exception is when there are multiple keys with exactly the same 64-bit
+ *    hash value. These are always returned together, since they correspond to
+ *    the same cursor value.
```

Comment:
There can be 64 bit hash collisions, for which storing the duplicates is not yet implemented. (The idea is to just store all the entries in level 14 (is it?) in a compact array and use one of the bitmaps for its length.)

Reply:
OK - so you're planning to diverge from the HAMT strategy of generating a 2nd hash to add more bits.

In this case, for iteration, maybe consider keeping the `key` and/or index and/or next-key pointer in the iterator.  Then, when the iterator hits such a list you scan until you hit the key you last iterated on.  This would avoid the weird pattern where an iterator can return multiple items in this rare case.  It's simpler for users if the iterator only returns 1 item each time.

If you're worried about the values changing/moving during iteration, maybe you could clone the list of keys into the iterator at this point, then, for each key, check that it still exists in the map before returning it.

Reply:
Good ideas. Yes, using a second hash function was the initial idea, but I think just storing the collisions is fine and simpler. I don't have a strong preference though.

I don't think the scan cursor should be more than 64 bits (maybe only 32 bits for 32-bit builds?) so SCAN would need to sometimes return more than the requested number of elements. An iterator should of course only return one element each time though, so it would need to store the duplicate index or a clone of the duplicates list in the iterator, yes.